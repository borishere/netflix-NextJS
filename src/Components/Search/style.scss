.search-wrap {
  margin: 50px 100px 100px;

  .search-title {
    font-size: 40px;
    font-weight: 300;
    margin-bottom: 40px;
  }

  .search-el {
    display: flex;
    flex-direction: row;
  }

  input {
    background: rgba(50, 50, 50, 0.8);
    mix-blend-mode: normal;
    opacity: 0.7;
    border-radius: 4px;
    font-family: Montserrat;
    font-size: 20px;
    font-weight: 400;
    padding: 17px 19px;
    border: 0;
    color: #fff;
    width: 100%;
  }

  .search-btn {
    background: #f65261;
    border-radius: 4px;
    font-family: Montserrat;
    font-size: 20px;
    font-weight: 500;
    padding: 18px 73px;
    margin-left: 14px;
    border: 0;
    color: #fff;
    cursor: pointer;
  }
}
