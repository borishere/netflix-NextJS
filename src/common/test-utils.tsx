import React, { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ModalContext } from '../Context/ModalContext';
import { Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../app/store';
import { createMemoryHistory } from 'history';
import { moviesApi } from '../services/movies';
import { setupListeners } from '@reduxjs/toolkit/dist/query';
import moviesReducer from '../app/moviesSlice';
import { configureStore } from '@reduxjs/toolkit';

const history = createMemoryHistory();


const AllTheProviders: FC = ({ children }) => {
  return (
    <Router location='/' navigator={history}>
      <Provider store={store}>
        <ModalContext.Provider value={{}}>
          {children}
        </ModalContext.Provider>
      </Provider>
    </Router>
  );
};

function renderWithStore(
  ui: ReactElement,
  {
    store = configureStore({
      reducer: {
        [moviesApi.reducerPath]: moviesApi.reducer,
        movies: moviesReducer

      },
      middleware: (getDefaultMiddleware: any) =>
        getDefaultMiddleware().concat(moviesApi.middleware)
    }),
    ...renderOptions
  } = {}
) {
  setupListeners(store.dispatch);

  function Wrapper({ children }: any) {
    const history = createMemoryHistory();

    return (
      <Provider store={store}>
        <Router location='/' navigator={history}>
          {children}
        </Router>
      </Provider>
    );
  }

  store.dispatch(moviesApi.util.resetApiState());

  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

const root = document.createElement('div');
root.id = 'root';

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, container: document.body.appendChild(root), ...options });

export * from '@testing-library/react';
export { customRender, renderWithStore };