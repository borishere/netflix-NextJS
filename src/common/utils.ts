export function addParamToExistsSearchParams(key: string, value: string): URLSearchParams {
  const params = new URLSearchParams(window.location.search);

  params.set(key, value);

  return params;
}

export function deleteParamFromExistsSearchParams(value: string): URLSearchParams {
  const params = new URLSearchParams(window.location.search);

  params.delete(value);

  return params;
}

export const BASE_API_URL: string = 'http://localhost:4000/';