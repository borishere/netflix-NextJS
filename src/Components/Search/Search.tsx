import React from 'react';
import { KeyboardEvent, useEffect, useState } from 'react';
import { useParams, useNavigate } from 'react-router-dom';
import './style.scss';

export const SearchForm = () => {
  const [value, setValue] = useState<string>('');
  const navigate = useNavigate();

  const { searchQuery } = useParams();

  useEffect(() => {
    if (searchQuery) {
      setValue(searchQuery);
    }
  }, [searchQuery]);

  const searchHandler = () => {
    navigate(`search/${value}`);
  };

  const onKeyDown = (e: KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter') {
      searchHandler();
    }
  };

  return (
    <div className='search-wrap'>
      <div className='search-title'>FIND YOUR MOVIE</div>
      <div className='search-el'>
        <input
          className='search-input'
          value={value}
          onChange={(e) => setValue(e.target.value)}
          placeholder='What do you want to watch?'
          onKeyDown={(e) => onKeyDown(e)}
        />
        <button
          className='search-btn'
          onClick={searchHandler}
          aria-label='search'
        >
          SEARCH
        </button>
      </div>
    </div>
  );
};
