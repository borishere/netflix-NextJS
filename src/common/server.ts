import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { BASE_API_URL } from './utils';
import mockData from './__mocks__/movies.json';

let movies: any[] = Array.from(mockData);

export const server = setupServer(
  rest.get(`${BASE_API_URL}movies`, (req, res, ctx) => {
    return res(ctx.json(movies.slice(0, 2)));
  }),

  rest.get(`${BASE_API_URL}movies:id`, (req, res, ctx) => {
    const { id } = req.params;
    const movie = movies.find((movie) => movie.id === id);

    return res(ctx.json(movie));
  }),

  rest.post(`${BASE_API_URL}movies`, (req, res, ctx) => {
    movies.push(req.body);

    return res(ctx.status(200));
  }),

  rest.put(`${BASE_API_URL}movies`, (req, res, ctx) => {
    const requestMovie: any = req.body;

    const movieIndex = movies.findIndex((m) => m.id === requestMovie.id);

    if (movieIndex < 0) {
      return res(ctx.status(404));
    }

    movies[movieIndex] = requestMovie;

    return res(ctx.status(200));
  }),

  rest.delete(`${BASE_API_URL}movies/:id`, (req, res, ctx) => {
    const { id } = req.params;

    movies = movies.filter((m) => m.id !== id);

    return res(ctx.status(200));
  })
);
