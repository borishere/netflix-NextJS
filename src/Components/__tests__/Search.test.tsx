import React from 'react';
import { screen, fireEvent, customRender } from '../../common/test-utils';
import { SearchForm } from '../Search/Search';
import '@testing-library/jest-dom';

test('Search', () => {
  const { asFragment } = customRender(<SearchForm />);
  expect(asFragment()).toMatchSnapshot();

  const input: any = screen.getByPlaceholderText('What do you want to watch?');
  expect(input).toBeInTheDocument();

  fireEvent.change(input, { target: { value: 'titanic' } });
  expect(input.value).toBe('titanic');
});
