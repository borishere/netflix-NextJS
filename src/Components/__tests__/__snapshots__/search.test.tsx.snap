// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search 1`] = `
<DocumentFragment>
  <div
    class="search-wrap"
  >
    <div
      class="search-title"
    >
      FIND YOUR MOVIE
    </div>
    <div
      class="search-el"
    >
      <input
        class="search-input"
        placeholder="What do you want to watch?"
        value=""
      />
      <button
        aria-label="search"
        class="search-btn"
      >
        SEARCH
      </button>
    </div>
  </div>
</DocumentFragment>
`;
